### PR TITLE
fix: correct floating point ChunkHasher

### LINKED ISSUE #3768: Floating point comparisons are inconsistent with hashCode
Floating point comparisons are treating `-0.0` and `0.0` as equal, which is inconsistent with `Double.hashCode` and `Float.hashCode`. This also prevents using `doubleToLongBits` or `floatToIntBits` for NaN-respecting equality testing.

### PR DESCRIPTION
**Pull request by devinrsmith**
This adds documentation and a new hashCode method to the `io.deephaven.util.compare` package that is consistent with Deephaven equality. Specifically, this ensures that the floating points values -0.0 and 0.0 hash to the same value.

Testing was added around aggregation keys, join keys, aggregations, min/max formulas, and sort results. Of note is that unique and distinct aggregations currently rely on `io.deephaven.chunk.WritableChunk#sort()`, which treats -0.0 < 0.0. This effects the encounter order returned by these aggregations, and could be seen as inconsistent with our stated goals of treating -0.0 == 0.0. This does not affect the consistency of the sort operation. It is hard to be confident that the testing coverage around this issue is fully complete; that said, this does make us much more consistent in the tested operations wrt Deephaven equality.

This is a prerequisite for #5605

Fixes #3768

### CODE REVIEW

File: Util/src/main/java/io/deephaven/util/compare/BooleanComparisons.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
+//
+// Copyright (c) 2016-2024 Deephaven Data Labs and Patent Pending
+//
+package io.deephaven.util.compare;
+
+public class BooleanComparisons {
```
Comment (rcaudy) [inline Util/src/main/java/io/deephaven/util/compare/BooleanComparisons.java:6]: Why do we need this? We don't ever allow a column to contain primitive booleans.

File: engine/chunk/src/main/java/io/deephaven/chunk/util/hashing/ObjectChunkIdentityEquals.java [hunk 2]
```diff
@@ -250,7 +251,7 @@ public void andEqualPairs(IntChunk<ChunkPositions> chunkPositionsToCheckForEqual
 
     // region eq
     static private boolean eq(Object lhs, Object rhs) {
-        return lhs == rhs;
+        return ObjectComparisons.eq(lhs, rhs);
```
Comment (rcaudy) [inline engine/chunk/src/main/java/io/deephaven/chunk/util/hashing/ObjectChunkIdentityEquals.java:254]: This is wrong.

File: engine/table/src/test/java/io/deephaven/engine/table/impl/QueryTableNegativeZeroTest.java [hunk 3]
```diff
@@ -0,0 +1,459 @@
...
+        final float result = table.getColumnSource(column, float.class).getFloat(key);
+        TestCase.assertEquals(Float.floatToIntBits(expectedExact), Float.floatToIntBits(result));
+    }
+
+    private static void doubleToBitsEquals(Table table, String column, long key, double expectedExact) {
+        final double result = table.getColumnSource(column, double.class).getDouble(key);
+        TestCase.assertEquals(Double.doubleToLongBits(expectedExact), Double.doubleToLongBits(result));
+    }
+
+    private static class CreateAggSpecsDhc3768 implements AggSpec.Visitor {
```
Comment (rcaudy) [inline engine/table/src/test/java/io/deephaven/engine/table/impl/QueryTableNegativeZeroTest.java:317]: This is overkill; hashing is consistent for all aggs. I would consider testing single column, and multi-column.

File: engine/table/src/test/java/io/deephaven/engine/table/impl/QueryTableNegativeZeroTest.java [hunk 4]
```diff
@@ -0,0 +1,459 @@
...
+            }
+            floatToBitsEquals(out, "KeyNeg", key, -0.0f);
+            floatToBitsEquals(out, "KeyPos", key, 0.0f);
+            floatToBitsEquals(out, "Neg", key, -0.0f);
+            floatToBitsEquals(out, "Pos", key, 0.0f);
+        }
+    }
+
+    @Test
+    public void testDhc3768_double_join_key() {
```
Comment (rcaudy) [inline engine/table/src/test/java/io/deephaven/engine/table/impl/QueryTableNegativeZeroTest.java:99]: rangeJoin and multiJoin?

File: engine/table/src/test/java/io/deephaven/engine/table/impl/QueryTableNegativeZeroTest.java [hunk 5]
```diff
@@ -0,0 +1,459 @@
...
+            }
+            allAggregations.add(Aggregation.AggCount("Count"));
+            allAggregations.add(Aggregation.AggFirstRowKey("FirstRowKey"));
+            allAggregations.add(Aggregation.AggLastRowKey("LastRowKey"));
+            TestCase.assertEquals(1, xy.aggBy(allAggregations, "X").size());
+        }
+    }
+
+    @Test
+    public void testDhc3768_min_max_result() {
```
Comment (rcaudy) [inline engine/table/src/test/java/io/deephaven/engine/table/impl/QueryTableNegativeZeroTest.java:170]: List of aggs to test:
* min (looking to preserve encounter order)
* max (looking to preserve encounter order)
* percentile
* median
* count distinct
* distinct
* unique
* sorted first
* sorted last

**Review by rcaudy — CHANGES_REQUESTED** [review]

File: Util/src/main/java/io/deephaven/util/compare/BooleanComparisons.java (hunk 1 above)
Comment (devinrsmith) [reply Util/src/main/java/io/deephaven/util/compare/BooleanComparisons.java:6]: This allows us to not have to special case some boolean code generation cases; namely, `BooleanChunkEquals` and `BooleanChunkHasher` will now call this class (just like the other ChunkEquals / ChunkHasher impls). Maybe we can make the argument that `BooleanComparisons` should only have `eq` and `hashCode`, and not `compare`, `gt`, etc...?

File: engine/chunk/src/main/java/io/deephaven/chunk/util/hashing/ObjectChunkIdentityEquals.java (hunk 2 above)
Comment (devinrsmith) [reply engine/chunk/src/main/java/io/deephaven/chunk/util/hashing/ObjectChunkIdentityEquals.java]: Good catch. Fixed. Surprised we didn't have any failing test.

File: engine/table/src/test/java/io/deephaven/engine/table/impl/QueryTableNegativeZeroTest.java (hunk 4 above)
Comment (devinrsmith) [reply engine/table/src/test/java/io/deephaven/engine/table/impl/QueryTableNegativeZeroTest.java]: Added multijoin. Spent a little time on rangeJoin, didn't quite grok the correct incantations.

File: engine/table/src/test/java/io/deephaven/engine/table/impl/QueryTableNegativeZeroTest.java (hunk 5 above)
Comment (devinrsmith) [reply engine/table/src/test/java/io/deephaven/engine/table/impl/QueryTableNegativeZeroTest.java]: I think I've got these bases covered now.

File: engine/table/src/test/java/io/deephaven/engine/table/impl/QueryTableFloatingPointZeroTest.java [hunk 6]
```diff
@@ -0,0 +1,599 @@
...
+    }
+
+    private static long oneKey(Table table) {
+        final long key;
+        try (final RowSet.Iterator iterator = table.getRowSet().iterator()) {
+            TestCase.assertTrue(iterator.hasNext());
+            key = iterator.nextLong();
+            TestCase.assertFalse(iterator.hasNext());
+        }
+        return key;
```
Comment (rcaudy) [inline engine/table/src/test/java/io/deephaven/engine/table/impl/QueryTableFloatingPointZeroTest.java:587]:
```suggestion
        TestCase.assertEquals(1, table.size());
        return table.getRowSet().firstRowKey();
```

File: engine/table/src/test/java/io/deephaven/engine/table/impl/QueryTableFloatingPointZeroTest.java [hunk 7]
```diff
@@ -0,0 +1,599 @@
...
+                t1.aj(t2, "Key1,KeyNeg>=KeyPos"),
+                t2.aj(t1, "Key1,KeyPos>=KeyNeg"),
+                t1.naturalJoin(t2, "KeyNeg=KeyPos,Key1"),
+                t2.naturalJoin(t1, "KeyPos=KeyNeg,Key1"),
+                t1.exactJoin(t2, "KeyNeg=KeyPos,Key1"),
+                t2.exactJoin(t1, "KeyPos=KeyNeg,Key1"),
+                t1.join(t2, "KeyNeg=KeyPos,Key1"),
+                t2.join(t1, "KeyPos=KeyNeg,Key1"),
+                t1.aj(t2, "KeyNeg=KeyPos,Key1"),
+                t2.aj(t1, "KeyPos=KeyNeg,Key1"))) {
```
Comment (rcaudy) [inline engine/table/src/test/java/io/deephaven/engine/table/impl/QueryTableFloatingPointZeroTest.java:115]: You actually need/want one more key column for AJ in each of your tests. There's only hashing on the first n-1 key columns, so the single-key tests are testing 0 hashed keys, and the multi-key tests are testing 1.

File: engine/table/src/test/java/io/deephaven/engine/table/impl/QueryTableFloatingPointZeroTest.java [hunk 8]
```diff
@@ -0,0 +1,599 @@
...
+            doubleToBitsEquals(out, "KeyNeg", key, -0.0);
+            doubleToBitsEquals(out, "KeyPos", key, 0.0);
+            doubleToBitsEquals(out, "Neg", key, -0.0);
+            doubleToBitsEquals(out, "Pos", key, 0.0);
+        }
+    }
+
+    @Test
+    public void testFloatJoinMultiKey() {
+        final Table t1 = newTable(intCol("Key1", 0), floatCol("KeyNeg", -0.0f), floatCol("Neg", -0.0f));
```
Comment (rcaudy) [inline engine/table/src/test/java/io/deephaven/engine/table/impl/QueryTableFloatingPointZeroTest.java:97]: I think you may want two fp keys. Mix and match double and float in both orders?

File: engine/table/src/test/java/io/deephaven/engine/table/impl/QueryTableFloatingPointZeroTest.java [hunk 9]
```diff
@@ -0,0 +1,599 @@
...
+    //
+    // @Test
+    // public void testRangeJoinFloatAggResults() {
+    //
+    // }
+    //
+    // @Test
+    // public void testRangeJoinDoubleAggResults() {
+    //
+    // }
```
Comment (rcaudy) [inline engine/table/src/test/java/io/deephaven/engine/table/impl/QueryTableFloatingPointZeroTest.java:213]: Delete me

File: engine/table/src/test/java/io/deephaven/engine/table/impl/QueryTableFloatingPointZeroTest.java [hunk 10]
```diff
@@ -0,0 +1,599 @@
...
+    @Test
+    public void testDoubleMultiJoin() {
+        // preserves Key from first table
+        {
+            final Table t1 = newTable(doubleCol("Key", -0.0), intCol("S1", 1));
+            final Table t2 = newTable(doubleCol("Key", 0.0), intCol("S2", 2));
+            final Table t3 = newTable(doubleCol("Key", -0.0), intCol("S3", 3));
+            final Table out = MultiJoinFactory.of("Key", t1, t2, t3).table();
+            final long key = oneKey(out);
+            doubleToBitsEquals(out, "Key", key, -0.0);
```
Comment (rcaudy) [inline engine/table/src/test/java/io/deephaven/engine/table/impl/QueryTableFloatingPointZeroTest.java:183]: Maybe test value columns, too?

File: engine/table/src/test/java/io/deephaven/engine/table/impl/QueryTableFloatingPointZeroTest.java [hunk 11]
```diff
@@ -0,0 +1,599 @@
...
+        }
+        {
+            final Table xy = newTable(floatCol("X", 0.0f, -0.0f), intCol("Y", 0, 1));
+            final Table grouped = xy.groupBy("X");
+            floatToBitsEquals(grouped, "X", oneKey(grouped), 0.0f);
+        }
+    }
+
+    @Test
+    public void testDoubleAggKey() {
```
Comment (rcaudy) [inline engine/table/src/test/java/io/deephaven/engine/table/impl/QueryTableFloatingPointZeroTest.java:230]: I think you should include multi-key tests for aggs, too.

File: engine/table/src/test/java/io/deephaven/engine/table/impl/QueryTableFloatingPointZeroTest.java [hunk 12]
```diff
@@ -0,0 +1,599 @@
...
+                    posFirst.groupBy().view("X=min(X)"),
+                    posFirst.groupBy().view("X=max(X)"))) {
+                doubleToBitsEquals(table, "X", oneKey(table), 0.0);
+            }
+        }
+    }
+
+    @Test
+    @Ignore("Unstable sort?")
+    public void testFloatMedianAggResult() {
```
Comment (rcaudy) [inline engine/table/src/test/java/io/deephaven/engine/table/impl/QueryTableFloatingPointZeroTest.java:293]: This is probably less well-defined than we'd hope. I bet that the SSM bucket gets created for -0.0 or 0.0, whichever is encountered first, and we'll keep using that "forever".

File: engine/table/src/test/java/io/deephaven/engine/table/impl/QueryTableFloatingPointZeroTest.java (hunk 8 above)
Comment (devinrsmith) [reply engine/table/src/test/java/io/deephaven/engine/table/impl/QueryTableFloatingPointZeroTest.java:97]: hoping a single float key in a variety of positions is good enough.

**Review by rcaudy — APPROVED** [review]

**Pull request merged**